contracts/services/FaucetService.sol

```diff
@@ -0,0 +1,128 @@
+
+    string constant public FLAVOR = "FaucetServiceV1";
+    string constant public VERSION = "1.0.0";
+
+    function flavor() external pure override returns (string memory) { return FLAVOR; }
+    function version() external pure override returns (string memory) { return VERSION; }
+
+    //------------------------
+    //  STATE
+    //------------------------
+    address public masterOwner;
+    mapping (address => bool) public allowedCallers;
+    uint256 public rewardPerApprove;
+    uint256 public balanceThresholdForReward;
+
+    //------------------------
+    //  EVENTS
+    //------------------------
+    event UpdateCallerStatus(address indexed caller, address indexed approver, bool approved, uint256 timestamp);
+    event WalletFunded(address indexed caller, address wallet, uint256 reward, uint256 timestamp);
+    event UpdateRewardAmount(address indexed caller, uint256 oldAmount, uint256 newAmount, uint256 timestamp);
+    event UpdateBalanceThresholdForReward(address indexed caller, uint256 oldThreshold, uint256 newThreshold, uint256 timestamp);
+    event OwnershipChanged(address indexed oldOwner, address indexed newOwner, uint256 timestamp);
+    event Received(address indexed sender, uint256 amount, uint256 timestamp);
+    event Released(address indexed receiver, uint256 amount, uint256 timestamp);
+
+    //------------------------
+    //  CONSTRUCTOR
+    //------------------------
+    constructor(address _masterOwner, address[] memory _callers, uint256 _rewardPerApprove, uint256 _balanceThresholdForReward) {
+        require(_masterOwner != address(0), "FaucetService: invalid master owner");
+        require(_rewardPerApprove > 0, "FaucetService: reward per approve must be positive");
+        require(_balanceThresholdForReward >= 0, "FaucetService: balance threshold for reward must not be negative");
+        
+        for (uint i = 0; i < _callers.length; i++) {
+            require(_callers[i] != address(0), "FaucetService: invalid caller address");
+            allowedCallers[_callers[i]] = true;
+        }
+
+        masterOwner = _masterOwner;
+        allowedCallers[masterOwner] = true;
+        rewardPerApprove = _rewardPerApprove;
+        balanceThresholdForReward = _balanceThresholdForReward;
+    }
+
+    //------------------------
+    //  MODIFIERS
+    //------------------------
+    modifier isMasterOwner {
+        require(msg.sender == masterOwner, "FaucetService: not master owner");
+        _;
+    }
+
+    modifier isAllowed {
+        require(
+            msg.sender == masterOwner || allowedCallers[msg.sender],
+            "FaucetService: not allowed to call function"
+        );
+        _;
+    }
+
+    //------------------------
+    //  STATE CHANGE FUNCTIONS
+    //------------------------
+    function faucet(address payable[] calldata _wallets) external override isAllowed {
+        require(rewardPerApprove > 0, "FaucetService: reward per approve must be positive");
```

Comment:
True, it cannot be negative. However, it can be zero and we want the reward to be always be > 0, otherwise there is no point in having a faucet contract, unless we want to temporarily have no-op calls but that would only waste gas. I will remove the check here as this is already checked in the constructor and `updateRewardAmount` but the non-zero check will remain. For `balanceThresholdForReward` however, all checks can be removed because it is `uint` and it's okay if that is set to 0.